pinecone/index_connection.go

```diff
@@ -0,0 +1,358 @@
+	IncludeMetadata bool
+	SparseValues    *SparseValues
+}
+
+type QueryVectorsResponse struct {
+	Matches   []*ScoredVector
+	Namespace string
+	Usage     *Usage
+}
+
+func (idx *IndexConnection) QueryByVector(ctx *context.Context, in *QueryByVectorRequest) (*QueryVectorsResponse, error) {
+	req := &data.QueryRequest{
+		Namespace:       in.Namespace,
+		TopK:            in.TopK,
+		Filter:          in.Filter,
+		IncludeValues:   in.IncludeValues,
+		IncludeMetadata: in.IncludeMetadata,
+		Vector:          in.Vector,
+		SparseVector:    sparseValToGrpc(in.SparseValues),
+	}
+
+	return idx.query(ctx, req)
+}
+
+type QueryByIdRequest struct {
+	Id              string
+	Namespace       string
+	TopK            uint32
+	Filter          *Filter
+	IncludeValues   bool
+	IncludeMetadata bool
+	SparseValues    *SparseValues
+}
+
+func (idx *IndexConnection) QueryById(ctx *context.Context, in *QueryByIdRequest) (*QueryVectorsResponse, error) {
+	req := &data.QueryRequest{
+		Id:              in.Id,
+		Namespace:       in.Namespace,
+		TopK:            in.TopK,
+		Filter:          in.Filter,
+		IncludeValues:   in.IncludeValues,
+		IncludeMetadata: in.IncludeMetadata,
+		SparseVector:    sparseValToGrpc(in.SparseValues),
+	}
+
+	return idx.query(ctx, req)
+}
+
+type DeleteVectorsRequest struct {
+	Namespace string
+	Ids       []string
+	Filter    *Filter
+	DeleteAll bool
+}
+
+func (idx *IndexConnection) DeleteVectors(ctx *context.Context, in *DeleteVectorsRequest) error {
+	req := data.DeleteRequest{
+		Ids:       in.Ids,
+		DeleteAll: in.DeleteAll,
+		Namespace: in.Namespace,
+		Filter:    in.Filter,
+	}
+
+	_, err := (*idx.dataClient).Delete(idx.akCtx(*ctx), &req)
+	return err
+}
+
+type UpdateVectorRequest struct {
+	Id           string
+	Values       []float32
+	SparseValues *SparseValues
+	Metadata     *Metadata
+	Namespace    string
+}
+
+func (idx *IndexConnection) UpdateVector(ctx *context.Context, in *UpdateVectorRequest) error {
+	req := &data.UpdateRequest{
+		Id:           in.Id,
+		Values:       in.Values,
+		SparseValues: sparseValToGrpc(in.SparseValues),
+		SetMetadata:  in.Metadata,
+		Namespace:    in.Namespace,
+	}
+
+	_, err := (*idx.dataClient).Update(idx.akCtx(*ctx), req)
+	return err
+}
+
+type DescribeIndexStatsRequest struct {
+	Filter *Filter
+}
+
+type DescribeIndexStatsResponse struct {
+	Dimension        uint32
+	IndexFullness    float32
+	TotalVectorCount uint32
+	Namespaces       map[string]*NamespaceSummary
+}
+
+func (idx *IndexConnection) DescribeIndexStats(ctx *context.Context, in *DescribeIndexStatsRequest) (*DescribeIndexStatsResponse, error) {
+	req := &data.DescribeIndexStatsRequest{
+		Filter: in.Filter,
+	}
+	res, err := (*idx.dataClient).DescribeIndexStats(idx.akCtx(*ctx), req)
+	if err != nil {
+		return nil, err
+	}
+
+	namespaceSummaries := make(map[string]*NamespaceSummary)
+	for key, value := range res.Namespaces {
+		namespaceSummaries[key] = &NamespaceSummary{
+			VectorCount: value.VectorCount,
+		}
+	}
```

Comment:
**question:** We do some mapping before returning responses for `FetchVectors`, `ListVectors`, `DescribeIndexStats`, and `query`, and I suppose I'm mainly curious why that is.

Reply:
mainly to not expose the `internal` struct types from the generated code. This allows us to have a clean and well documented interface in the `pinecone` package without the user of the SDK having to dig into internal code.